File: pt-4/164.maximum-gap.go
package leetcode

// MaximumGap164 uses bucket spacing from the pigeonhole principle. The maximum
// adjacent sorted gap must occur between non-empty buckets, so each bucket only
// stores its min and max.
//
// Time: O(n). Space: O(n).
func MaximumGap164(nums []int) int {
	n := len(nums)
	if n < 2 {
		return 0
	}
	lo, hi := nums[0], nums[0]
	for _, x := range nums {
		lo = minInt(lo, x)
		hi = maxInt(hi, x)
	}
	if lo == hi {
		return 0
	}
	size := maxInt(1, (hi-lo+n-2)/(n-1))
	count := (hi-lo)/size + 1
	bmin := make([]int, count)
	bmax := make([]int, count)
	used := make([]bool, count)
	for _, x := range nums {
		i := (x - lo) / size
		if !used[i] {
			bmin[i], bmax[i], used[i] = x, x, true
		} else {
			bmin[i] = minInt(bmin[i], x)
			bmax[i] = maxInt(bmax[i], x)
		}
	}
	best := 0
	prevSet := false
	prev := 0
	for i := 0; i < count; i++ {
		if !used[i] {
			continue
		}
		if prevSet {
			best = maxInt(best, bmin[i]-prev)
		}
		prev = bmax[i]
		prevSet = true
	}
	return best
}

File: pt-4/164.maximum-gap.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    int maximumGap(vector<int>& nums) {
        /*
        Approach: bucket by value range using the pigeonhole principle. The max
        gap cannot be inside a bucket when bucket size is chosen as ceil(range /
        (n-1)); it must be between the max of a non-empty bucket and the min of
        the next non-empty bucket.

        C++ notes: each Bucket stores whether it is used plus min/max values.
        Complexity: O(n) time, O(n) space.
        */
        int n = nums.size();
        if (n < 2) return 0;
        int lo = *min_element(nums.begin(), nums.end());
        int hi = *max_element(nums.begin(), nums.end());
        if (lo == hi) return 0;
        int size = max(1, (hi - lo + n - 2) / (n - 1));
        int count = (hi - lo) / size + 1;
        struct Bucket { bool used = false; int mn = INT_MAX; int mx = INT_MIN; };
        vector<Bucket> buckets(count);
        for (int num : nums) {
            int b = (num - lo) / size;
            buckets[b].used = true;
            buckets[b].mn = min(buckets[b].mn, num);
            buckets[b].mx = max(buckets[b].mx, num);
        }
        int best = 0, prevMax = lo;
        bool havePrev = false;
        for (const Bucket& bucket : buckets) {
            if (!bucket.used) continue;
            if (havePrev) best = max(best, bucket.mn - prevMax);
            prevMax = bucket.mx;
            havePrev = true;
        }
        return best;
    }
};
